google_cloud_automlops/orchestration/kfp/scaffold.py

```diff
@@ -106,7 +109,40 @@ def get_packages_to_install_command(func: Optional[Callable] = None,
     return ['sh', '-c', install_python_packages_script, src_code]
 
 
-def get_function_parameters(func: Callable) -> dict:
+def get_function_return_types(func: Callable) -> list:
+    """Returns a formatted list of function return types.
+
+    Args:
+        func: The python function to create a component from. The function
+            can optionally have type annotations for its return values.
+    Returns:
+        list: return value list with types converted to kubeflow spec.
+    Raises:
+        Exception: If return type is provided and not a NamedTuple.
+    """
+    annotation = inspect.signature(func).return_annotation
+    if maybe_strip_optional_from_annotation(annotation) is not annotation:
+        raise TypeError('Return type cannot be Optional.')
+
+    # No annotations provided
+    # pylint: disable=protected-access
+    if annotation == inspect._empty:
+        return None
+
+    if not (hasattr(annotation,'__annotations__') and isinstance(annotation.__annotations__, dict)):
+        raise TypeError(f'''Return type hint for function "{func.__name__}" must be a NamedTuple.''')
+
+    outputs = []
+    for name, type_ in annotation.__annotations__.items():
+        metadata = {}
+        metadata['name'] = name
+        metadata['type'] = type_
+        metadata['description'] = None
```

Comment:
Small nit: Can you add descriptions from the docstring, if they exist? For congruency's sake - code for this is in `get_function_parameters`

Reply:
Unfortunately there isn't really a well established way to document individual return values for a NamedTuple. We could add the following incase someone really wanted this capability:


```
import docstring_parser
import inspect

def func():
    """
    Function that returns a named tuple with outputs.

    Returns:
        NamedTuple:
            a: Description of a
            b: Description of b
    """

parsed_docstring = docstring_parser.parse(inspect.getdoc(func))

# Grab full description
description = parsed_docstring.returns.description
print(description)
#  a: Description of a
#  b: Description of b


# Then manually parse out the type for each
descriptions = [line.split(":") for line in description.split("\n")]
doc_dict = {parts[0].strip() : ":".join(parts[1:]).strip() for parts in descriptions}
print(doc_dict)
# {'a': 'Description of a', 'b': 'Description of b'}
```

Reply:
Fair point, there does seem to be a couple ways people tend to describe return values in the docstring. Let's go ahead and leave it as None for now